### PR TITLE
Persist bearer token details across page reloads

### LINKED ISSUE #72: Oauth sign-in fails in Chrome when third party cookies blocked
Linked to #30 

The oauth flow works and the access_token is returned in the URL fragment however that redirect response tries to set a cookie as well. I assume this cookie set failure causes the code / a promise to fail even though we have all the information we need to authenticate the user. 
Steps to recreate:
1. Activate block third party cookies in chrome
0. try to login to shakespearesworld.org and check the response from panoptes signin oauth flow.

Happy to help debug this as well

### PR DESCRIPTION
**Pull request by eatyourgreens**
Use window session storage to store the current bearer token across page reloads, in case 3rd party cookies are blocked.

I think this will fix #72, where sign-in does not work on sites like antislaverymanuscripts.org if 3rd party cookies are disabled.

It does not fix token refreshes, for long sessions, being blocked in that case.

### CODE REVIEW

File: lib/oauth.js [hunk 1]
```diff
@@ -152,9 +153,14 @@ module.exports = new Model({
   },
 
   _checkForPanoptesSession: function() {
+    var sessionTokenDetails = sessionStorage.getItem(LOCAL_STORAGE_PREFIX + 'tokenDetails');
```
Comment (eatyourgreens) [inline lib/oauth.js]: missing JSON.parse here.

**Comment by eatyourgreens** [comment]
Reading up on [sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage), keys are stored for the page session, not browser session, so this code won't preserve the token across tabs or windows.

**Comment by simoneduca** [comment]
I wonder if this might introduce changes that confuse/frustrate volunteers. What if they accidentally close their current window and lose work?

**Comment by eatyourgreens** [comment]
I don't follow. At the moment they can't log in at all.

**Comment by eatyourgreens** [comment]
I've verified on a local copy of Shakespeare's World that this fixes zooniverse/shakespeares_world#347. I can log in when 3rd party cookies are blocked. I'm not sure if login persists across tabs.

**Comment by simoneduca** [comment]
I see that being able to login is critical, but how many people have 3rd parties cookies disabled? Session not persisting in new tabs would affects lots of users I think. Ideally we would like both, no?

**Comment by eatyourgreens** [comment]
Having read the [MDN description of `sessionStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage), I'm not completely convinced it's appropriate here.

> I see that being able to login is critical, but how many people have 3rd parties cookies disabled? 

I think this misses the point. @camallen pointed out that the OAuth flow we use means that shakespearesworld.org should not be reading a 3rd party cookie from panoptes.zooniverse.org to log in. It should work with the token that's passed back in the redirect URL from the Zooniverse sign-in page.

If I remember correctly, the original bug report was Safari on OSX, so that may have 3rd party cookies disabled by default. They'll also be blocked by anyone who uses an ad-blocker, or who blocks tracking cookies for privacy reasons.

**Comment by simoneduca** [comment]
>Having read the MDN description of sessionStorage, I'm not completely convinced it's appropriate here.
What alternatives to we have then?

**Comment by eatyourgreens** [comment]
Well `localStorage`, but local storage never expires, so code will probably be needed to manually remove expired tokens when a login session expires. I haven't looked into the oauth code in enough detail to see if it already does this.

**Comment by eatyourgreens** [comment]
Another alternative would be to rewrite the API for the oauth module, so that the client app (Shakespeare's World/ASM/Intro2Astro) is responsible for tracking expired tokens and issuing refresh requests. I'm not sure about making changes that break live sites.

**Comment by simoneduca** [comment]
>Well localStorage, but local storage never expires, so code will probably be needed to manually remove expired tokens when a login session expires. I haven't looked into the oauth code in enough detail to see if it already does this.

This is meant to refresh the token before it actually expires https://github.com/zooniverse/panoptes-javascript-client/blob/master/lib/oauth.js#L242

**Comment by eatyourgreens** [comment]
Yes, but that uses an iframe to refresh the token, which is why this 3rd-party domain bug exists in the first place.

**Comment by simoneduca** [comment]
Btw why do you think sessionStorage isn't suitable here @eatyourgreens ?

**Comment by eatyourgreens** [comment]
@simoneduca The MDN docs imply that it uses page sessions, rather than browser sessions. See [my original comment](#issuecomment-361217990) in this thread.

**Comment by eatyourgreens** [comment]
Confirmed on a local copy of Shakespeare's World  in Chrome with 3rd party cookies disabled, and this version of the client installed. If I login, then cmd-click 'Get started' to open in a new tab, I'm not logged in on the transcription tab. Auth tokens don't persist across tabs with this PR.

If `sessionStorage` is replaced with `localStorage` in this PR, then the client will store, and try to use, expired tokens forever, because there's no code in place to check a token's validity before using it in `_handleBearerToken`. https://github.com/zooniverse/panoptes-javascript-client/blob/85bdc163aefaa3c84e9fac101408f590dfba8636/lib/oauth.js#L231-L240

**Comment by simoneduca** [comment]
I'd be happy to pair up to add the necessary code. But I'm not sure using `localStorage` is right either.

File: lib/oauth.js [hunk 2]
```diff
@@ -179,7 +185,7 @@ module.exports = new Model({
             }
           } catch (error) {
             if (error.name === 'SecurityError') {
-              error = new Error('No existing Panoptes session found');
+              console.warn('No existing Panoptes session found');
```
Comment (simoneduca) [inline lib/oauth.js]: why not logging the actual `error` as well?

Comment (eatyourgreens) [reply lib/oauth.js]: That happens on the next line, when the promise rejects.

**Comment by eatyourgreens** [comment]
> But I'm not sure using localStorage is right either.

@simoneduca can you expand on why using local storage isn't right?

**Comment by camallen** [comment]
By storing the token we increase the attack surface for clickjacking and XSS attacks, as described in the issue you linked above @simoneduca. This is one consideration when implementing the login flows on our apps. 

Also the bearer token is a JWT (https://jwt.io/), it is an encoded & signed payload that includes the user details and the expiry of the token. Check the debugger at the link above and paste in your JWT from the panoptes server. It should include details like:
```
{
  "data": {
    "login": "username",
    "dname": "username",
    ...
  },
  "exp": 1517312782,
   ...
}
```

So we can use the token expiration time to modify the renewal behaviour instead of relying on the `setTimeout` callback. Also all our auth tokens are JWT so we could create some shared code to deal with token renewal for PFE, implicit grant oauth apps, etc instead of dealing with them all differently. 

The implicit grant (oauth.js) differs from the way PFE handles authentication. Mostly that it authenticates on another site (signin/panoptes.zooniverse.org) and noting that the implicit flow doesn't return a refresh token (used to get a new JWT when the current one expires in PFE). Implicit flows used iframes to request a new token but rely on the session cookie with panoptes to verify they are allowed, blocking these cookies (3rd party) breaks the flow with iframes as their is no session cookie passed so panoptes thinks there is no session :( 

We allow PFE to use the credentials flow (submit user creds directly) to gain access to JWT token & refresh tokens, thus zooniverse.org doesn't suffer from the 3rd party cookie issue, though it all runs on the same domain so it would be ok if we did. 

We could change our implicit apps to use the credential flows as they get refresh tokens this way and thus avoid the use of iframes / consolidate on one flow / code base. I can't remember the reason why we moved to implicit for custom apps, perhaps to simplify / standardize the authentication code  ? 

Alternatively we can build out he implicit flows with detection / fallback for 3rd party cookie blocking. As the first login flow uses full page redirects to return the JWT not an iframe, then if i have a JWT but i can't refresh it using iframe them i'm most likely blocking 3rd party cookies. Noting that Jim has debugged this and it appears that the sigin bug i described in the original issue is due to the JWT returned in the hash fragment is ignored and we request one again from an iframe :sadpanda:. 

Noting that the UX on blocked 3rd party cookies is poor, notify the user is being logged out, save work? and force them to login again (or whitelist the blocked domain?) through page redirects. Though i am not sure how best to avoid this for blocked sites, maybe just warn they will be logged out in x seconds due to blocking our login cookies...?

I hope the above is useful, happy to clarify the oauth flows / options we have as it can be quite confusing.

**Comment by eatyourgreens** [comment]
Any thoughts on the code in this PR? Before I do any more work on it.

File: lib/oauth.js [hunk 3]
```diff
@@ -62,10 +65,11 @@ module.exports = new Model({
...
-      if (checkUrlForToken(window.location.href)) {
+      if (checkUrlForToken(window.location.hash)) {
         console.log('Token found in URL');
-        var tokenDetails = parseUrl(window.location.href);
+        var tokenDetails = parseUrl(window.location.hash);
         this._handleBearerToken(tokenDetails);
+        SESSION_STORAGE.setItem(LOCAL_STORAGE_PREFIX + 'tokenDetails', JSON.stringify(tokenDetails));
 
         // And redirect to the desired page
         var url = ls.get(LOCAL_STORAGE_PREFIX + 'redirectUri');
```
Comment (camallen) [inline lib/oauth.js]: Is the redirect straight after this local storage fetch for a saved page location that a user want's to access? E.g. `/#!/transcribe` ? I'm not sure why we redirect again after the authentication server round trip? Perhaps the `redirectUri` name is confusing as it's also used in the oauth authentication flow..

**Comment by eatyourgreens** [comment]
> The authentication code flows seem a little confused but i think this will works fine to solve the borked can't login issue.

Tell me about it! The bug turns out to be really simple, but understanding the code had me going around in circles for a bit. I think the code could be cleaned up a lot, but that might be a long job and I wasn't sure how quickly a fix is needed for the inability to log in.

>However, I don't this this will renew a token correctly once it has expired as it'll gets the data from local storage and returns that if it is set. happy to be wrong but will this break long sessions on existing apps?

I'm using `sessionStorage`, so the token gets deleted when the window closes. Long, continuous sessions should, in theory, be killed by the javascript timeout that's set when you first log in but I haven't tested this.

File: lib/oauth.js (hunk 3 above)
Comment (eatyourgreens) [reply lib/oauth.js]: I'd assume so, from the comments in the code. I don't know why it requests a new URL from the server, rather than using `pushState` to update the history without reloading the page.

By the way, `location.assign()` can't be passed a relative URL. It must be passed a fully qualified URL. This is what is causing the sign-in bug, because the browser will throw away the bearer token when the page refreshes, and initialise the JS app from scratch.

My best guess as to why the code is written like this would be that this library doesn't know how the client app is set up, so doesn't know how to call a new page short of making a new page request. For example, would a React app need different code here from an Angular app and, if so, why should this library suddenly have to know about javascript app frameworks and how they navigate between views? A better oauth API might do the token initialisation, then leave it up to the calling app to decide how to navigate back to the original view and re-render the page.

**Comment by eatyourgreens** [comment]
A quick note on security:

>By storing the token we increase the attack surface for clickjacking and XSS attacks, as described in the issue you linked above @simoneduca. This is one consideration when implementing the login flows on our apps.

Window `Storage` objects can only be read on the domain that set them eg. javascript code loaded from zooniverse.org or facebook.com or wherever to a page on shakespearesworld.org could not read the stored token.

**Comment by camallen** [comment]
> I wasn't sure how quickly a fix is needed for the inability to log in.

It's been broken for a long time, it's not super urgent but something to fix and the current behaviour it's not great UX from us. 

>Long, continuous sessions should, in theory, be killed by the javascript timeout that's set when you first log in but I haven't tested this. 

This code is so entangled it confusing to get the flow right in my head...here is my take on this PR (please correct me where it's wrong). Assuming a blanks slate (no data in session store):
1. If i log in via the signin button, the changes in this PR will get my token from the URL hash fragment and stored them in local storage. Awesome i can now signin! 
2. At the same time it will schedule a callback to `refreshBearerToken` in 1hr 59 mins (2 hrs default validity from server).
3. Lets say I work on transcriptions for the 1hr 59 mins, the callback fires and calls `checkForPanoptesSession` which retrieves my current token details from session storage and by default will return the expiring / expired token and reset it in local storage and wire up the client auth headers to use this expired token and repeat steps 2-3. 

If what i believe is happening in 2-3 is correct in this PR change, then our sites will effectively log the user out after 1hr 59 mins but will not update the UI so to users they will appear logged in. All subsequent API interactions will be not logged in, new subjects fetch, classification submission, etc. Talk will work fine as it's a redirect to zooniverse.org so users will report they worked on a subject but it won't be recorded for them. 

This may be already happening if there is a delay between the `setTimeout` firing compounded by slow network / server not responding, etc. Also i'm not sure how the current code handles a missing server (gateway timeout or something similar).

**Comment by eatyourgreens** [comment]
You're right. Somewhere in the current code, it should be calling either `signOut()` or `_deleteBearerToken` before trying to refresh the token when the timeout expires. At that point, either the iframe works and refreshes the token, or the site just signs you out.

I don't think there's any way, with the API defined in this code, for the client app to be notified that you've been signed out. That would probably require a callback to be set when the app initialises. There might be an auth change event that Shakespeare's World could listen to, to detect when the user has been logged out.

Oddly, I'm glad I'm not the only one who finds this code hard to follow. I was starting to think I was just missing something here when I couldn't quite understand what was going on.

**Comment by eatyourgreens** [comment]
Should ` _refreshBearerToken` call `_deleteBearerToken()` before it checks for a new session? That would be a simple change.
https://github.com/zooniverse/panoptes-javascript-client/blob/85bdc163aefaa3c84e9fac101408f590dfba8636/lib/oauth.js#L242-L250

**Comment by eatyourgreens** [comment]
Once again we're swallowing errors and printing an innocuous message instead 😠 
https://github.com/zooniverse/panoptes-javascript-client/blob/85bdc163aefaa3c84e9fac101408f590dfba8636/lib/oauth.js#L247-L249

File: lib/oauth.js [hunk 4]
```diff
@@ -240,8 +249,10 @@ module.exports = new Model({
   },
 
   _refreshBearerToken: function() {
+    this._deleteBearerToken();
```
Comment (camallen) [inline lib/oauth.js]: If we unset the `_currentUserPromise` like https://github.com/zooniverse/panoptes-javascript-client/blob/85bdc163aefaa3c84e9fac101408f590dfba8636/lib/oauth.js#L122 we may be able to signal to the calling app that the authentication renewal failed. 

If we can get the the calling app react to this state we can update the UI. I don't think we can do this in SW / Annotate via [auth factory](https://github.com/zooniverse/shakespeares_world/blob/1d52a27215960b594893e69c40dc57807cf527cc/app/modules/auth/auth.factory.js#L15) without a code change.

**Comment by srallen** [comment]
This is working with edu API front end with 3rd party cookies disabled in chrome, though, it may need some UI work similar to the other apps to signal to re-sign in. Fixing the google sign in will be more involved because I forgot that it's basically impossible to test on staging because of this unresolved issue: zooniverse/Panoptes#2024

File: lib/oauth.js (hunk 4 above)
Comment (eatyourgreens) [reply lib/oauth.js]: That promise resolves when `checkCurrent()` resolves, as far as I can tell. Once a promise is resolved, further changes don't affect it. So, I think this line here cancels any calls to `checkCurrent()` that are in progress from the client app, by immediately resolving the session check with a null value.

**Comment by eatyourgreens** [comment]
One more small change, to report that error that's currently being swallowed, then I'm going to merge this and publish a minor version bump for the client.
